src/torchjd/autogram/_gramian_reverse_accumulator.py

```diff
@@ -0,0 +1,152 @@
+            >>>     output = model(input)
+            >>>     losses = criterion(output, target)
+            >>>
+            >>>     optimizer.zero_grad()
+            >>>     gramian = gramian_reverse_accumulator.compute_gramian(losses)
+            >>>     losses.backward(weighting(gramian))
+            >>>     optimizer.step()
+    """
+
+    def __init__(
+        self,
+        modules: Iterable[nn.Module],
+    ):
+        self._handles: list[TorchRemovableHandle] = []
+        self._gramian_accumulator = GramianAccumulator()
+        self._activable_hook_factory = ActivableHookFactory()
+        self._node_activator = NodeActivator()
+        self._target_edges = EdgeRegistry()
+
+        self._track_modules(modules)
+
+    def _track_modules(self, modules: Iterable[nn.Module]) -> None:
+        _modules = set(modules)
+
+        # Add module forward hooks to compute jacobians
+        for module in _modules:
+            if next(module.parameters(recurse=False), None) is None:
+                # Skip un-parameterized modules
+                continue
+
+            module_hook = self._activable_hook_factory.make_activable_hook(
+                ModuleHook(self._target_edges, self._gramian_accumulator, self._node_activator)
+            )
+            handle = module.register_forward_hook(module_hook)
+            self._handles.append(handle)
+
+    def deaugment_modules(self) -> None:
+        """
+        Used to de-augment the provided modules. It removes all the hooks that we have added.
+
+        Typical usage is:
+
+        >>> # Augment the model
+        >>> gramian_reverse_accumulator = GramianReverseAccumulator(model.modules())
+        >>>
+        >>>  # Use it
+        >>>  # ...
+        >>>
+        >>> # De-augment the model
+        >>> gramian_reverse_accumulator.deaugment_modules()
+        >>> # All hooks added by augment_model_for_iwrm have now been removed
+        """
+
+        for handle in self._handles:
+            handle.remove()
+
+    def compute_gramian(self, output: Tensor, grad_outputs: Tensor | None = None) -> Tensor:
+        """
+        Compute the Gramian of the Jacobian of `output` with respect the direct parameters of all
+        `modules`.
+
+        :param output: The vector to differentiate. Must have `ndim == 1`.
+        :param grad_outputs: The tangents for the differentiation. Default to a vector of 1s of the
+            same shape as `output`.
+        :returns: the Gramian of the Jacobian of `output` with respect to the direct parameters of
+            all `modules`
+        """
+
+        if output.ndim != 1:
+            raise ValueError(
+                "We currently support computing the Gramian with respect to vectors only."
+            )
+
+        if grad_outputs is None:
+            grad_outputs = torch.ones_like(output)
+
+        self._activable_hook_factory.deactivate()
+        self._node_activator.activate()
```

Comment:
It is worth noting that the two activation are always opposite of each other, so maybe we can factor.